### PR TITLE
MAINT: rename scipy.ndimage modules

### PR DESCRIPTION
**Pull request by AnirudhDagar**
#### Reference issue
This is from the discussion in [#14356 comment](https://github.com/scipy/scipy/pull/14356#issuecomment-884432633).
Also related to #14360.

#### What does this implement/fix?
This PR renames the `ndimage` modules from `<module.py>` to `<_module.py>` and also adds a new file named as `<module.py>` which is simply a wrapper to bring all private module functions under the old public looking namespace for backwards compatibility.

All this has to be done in anticipation of landing #14356 and later raising a PR to deprecate public looking private modules in `ndimage` as discussed in #14360 (similar to the pattern in #14419).

#### Additional information
Note: The huge diff has been explained [here](https://github.com/AnirudhDagar/scipy/pull/1) and is unavoidable here. It is okay in this situation as mentioned in the [#14356 comment](https://github.com/scipy/scipy/pull/14356#issuecomment-884242636).

I'm also expecting some failed linting tests (**Edit 1**: similar to #14419) which can be discussed/ignored.

I request the reviewers to review this commit by commit if possible (it might help to avoid the huge diff). Looking forward to the feedback and any questions/suggestions.

cc @rgommers @grlee77 @peterbell10

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -596,6 +596,21 @@ ignore_errors = True
 [mypy-scipy.ndimage.morphology]
 ignore_errors = True
```
Comment (AnirudhDagar) [inline mypy.ini:597]: I guess it will be okay to remove these now. Right?

Comment (rgommers) [reply mypy.ini:597]: I'm guessing an underscore should be added (`._morphology`) or Mypy will still be unhappy.

File: mypy.ini [hunk 2]
```diff
@@ -596,6 +596,21 @@ ignore_errors = True
...
+[mypy-scipy.ndimage._fourier]
+ignore_errors = True
+
+[mypy-scipy.ndimage._interpolation]
+ignore_errors = True
+
+[mypy-scipy.ndimage._measurements]
+ignore_errors = True
+
+[mypy-scipy.ndimage._morphology]
```
Comment (AnirudhDagar) [inline mypy.ini:611]: @rgommers I've added underscores here but did not remove the old ones yet. I guess it is okay to get rid of
LOC 584 to 597 (GitHub is not letting me select all these lines to comment for some reason.)

**Comment by AnirudhDagar** [comment]
From failing mypy tests [here](https://github.com/scipy/scipy/pull/14447/checks?check_run_id=3127617737).
> scipy/ndimage/tests/test_fourier.py:145: error: Module has no attribute "fourier_ellipsoid"  [attr-defined]
> scipy/ndimage/tests/test_fourier.py:146: error: Module has no attribute "fourier_gaussian"  [attr-defined]
> scipy/ndimage/tests/test_fourier.py:147: error: Module has no attribute "fourier_uniform"  [attr-defined]

Any suggestion on how to fix these? I'm not 100% sure about why they are caused. I also saw them while running mypy locally. Would be okay to use `# type: ignore` for these?

**Comment by rgommers** [comment]
> Any suggestion on how to fix these? I'm not 100% sure about why they are caused. I also saw them while running mypy locally. Would it be okay to use `# type: ignore` for these?

Yes, I'm seeing more of these locally - looks like a Mypy bug. It understand `ndimage.fourier_gaussian`, _except_ when it's inside a `pytest.mark.parametrize` statement. Adding ` type: ignore` seems fine. And when you do, it's good practice to put a comment above it explaining what the problem is. Ideally linking to a Mypy issue if one exists.

**Comment by AnirudhDagar** [comment]
> except when it's inside a pytest.mark.parametrize statement

Uhmm, but what about these two errors?:
`scipy/signal/_savitzky_golay.py:4: error: Module "scipy.ndimage" has no attribute "convolve1d" [attr-defined]`.
`scipy/signal/tests/test_savitzky_golay.py:6: error: Module "scipy.ndimage" has no attribute "convolve1d" [attr-defined]`

They do not involve any `pytest.mark.parameterize`. It seems I'm missing something here with my limited knowledge of mypy.

**Comment by rgommers** [comment]
Ah okay, it's something else. You should change these imports in `__init__.py` to use the `_module.py` versions:
```python
from .filters import *
from .fourier import *
from .interpolation import *
from .measurements import *
from .morphology import *
```

I have no idea why Mypy triggers on some star imports but not others. But I can reproduce the problem and it goes away with the above fix.

**Comment by AnirudhDagar** [comment]
> Ah okay, it's something else. You should change these imports in __init__.py to use the _module.py versions:

I don't think that's a good idea. This will break BC. If I do so, the whole purpose of adding the old named modules back is defeated.

Something like this will not work.
```python
In [1]: from scipy import ndimage
In [2]: ndimage.filters.convolve1d
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-919b64dba7c9> in <module>
----> 1 ndimage.filters.convolve1d

AttributeError: module 'scipy.ndimage' has no attribute 'filters'
```

I'll try to look deeply into what is happening with `[attr-defined]` mypy error.

**Comment by AnirudhDagar** [comment]
I'm able to avoid this mypy error with either of the two solutions:

1. Explicitly import all the functions inside `fourier.py` like the following

```python
from ._fourier import * # <- added this
from . import _fourier

__all__ = ['fourier_gaussian', 'fourier_uniform', 'fourier_ellipsoid',
           'fourier_shift']

def __dir__():
    return __all__

def __getattr__(name):
    if name not in __all__:
        raise AttributeError(
            f"scipy.ndimage.fourier has no attribute {name}.")

    return getattr(_fourier, name)
```



2. I get rid of `__getattr__` inside `fourier.py` and change a little how modules are imported.

```python
from ._filters import *

__all__ = ['fourier_gaussian', 'fourier_uniform', 'fourier_ellipsoid',
           'fourier_shift']

def __dir__():
    return __all__

def __getattr__(name):
    if name not in __all__:
        raise AttributeError(
            f"scipy.ndimage.fourier has no attribute {name}.")

    return locals()[name]() # <- make use of locals
```

How about taking approach 1? My guess is this will also solve the linting issue mentioned [here](https://github.com/scipy/scipy/pull/14419#issuecomment-884484969)

**Comment by grlee77** [comment]
I don't have any particular insight into the mypy, issue, but if you are importing using

```Python
from ._filters import *
```
and have also defined `__all__`, then do we even need to introduce `__getattr__` at all in this PR? Can't we add it later in the PR that will deal with the deprecation warnings?

**Comment by AnirudhDagar** [comment]
I agree with @grlee77. It is actually redundant here and should be added in the deprecation PR where it is actually needed.

**Comment by AnirudhDagar** [comment]
Found a similar problem mentioned in [#13699 comment](https://github.com/scipy/scipy/pull/13699#issuecomment-801993499) which was resolved by ignoring the error in this [commit](https://github.com/scipy/scipy/pull/13699/commits/cbb5573f11c61ac0da0d316e8f4d72dca01dcb37).

Maybe @mdhaber knows something. Please let me know if ignoring this mypy issue is the way to go ahead. Thanks in advance :))

**Comment by rgommers** [comment]
> I don't think that's a good idea. This will break BC. If I do so, the whole purpose of adding the old named modules back is defeated.

Ah I missed one little thing. It only breaks this (using `ndimage.module` without importing it explicitly):
```
import ndimage
ndimage.filters.fourier_gaussian
```
But that is easily fixed by:
```
from ._filters import *
import .filters
# and same for other imports
```
Do you agree that that doesn't break BC? If so, that seems easiest.

**Comment by AnirudhDagar** [comment]
Thanks! This sounds like a good option. Pushed the change.

**Comment by AnirudhDagar** [comment]
The lint failures here are the same as mentioned in [#14419 comment](https://github.com/scipy/scipy/pull/14419#issuecomment-884484969).
> Lint failure is because flake8 does not yet understand getattr. This failure will go away when the PR is merged; it can be ignored.

**Comment by rgommers** [comment]
Thanks @AnirudhDagar, this is looking pretty good now. We can't squash-merge because the `git mv` commit should be separately, but it looks like the rest should be in a single commit. Could you update this PR to contain two commits?

**Comment by AnirudhDagar** [comment]
Sure! I initially made individual commits to make sure it is easier for review. If it looks good I'll just squash everything except renaming commit into one and force push.

**Comment by rgommers** [comment]
One more issue left, xref gh-14456. The `__all__` dicts in `filters.py` and the other non-underscored modules need `# noqa: F822`. Let's get gh-14456 merged first, and then if you can amend the second commit, the lint failures here should go away.

**Comment by AnirudhDagar** [comment]
@rgommers I force pushed the changes, adding  `# noqa: F822` to `__all__`. I also had to add some noqa error comments in the `__init__.py` imports (saw them popup after running flake8 locally).

The current CI lint failures are not my changes. They happen to come from the unchanged part (the private modules). I suppose they are showing up in the PR as they are now in the diff (technically unchanged but github adds them to diff).

Should I fix them or leave them?

**Comment by rgommers** [comment]
They're all things like `E501 line too long  (80 > 79 characters)` - you can ignore those. It'd be a net negative to break those lines or add `# noqa: E501` to them (or at the very least, not worth spending any time on).

**Comment by AnirudhDagar** [comment]
Cool! If there aren't any other suggested changes for this PR, it is ready.

File: scipy/ndimage/_morphology.py [hunk 3]
```diff
@@ -1217,7 +1217,7 @@ def grey_erosion(input, size=None, footprint=None, structure=None,
     if size is None and footprint is None and structure is None:
         raise ValueError("size, footprint, or structure must be specified")
 
-    return filters._min_or_max_filter(input, size, footprint, structure,
+    return _filters._min_or_max_filter(input, size, footprint, structure,
                                       output, mode, cval, origin, 1)
```
Comment (peterbell10) [inline scipy/ndimage/_morphology.py:1221]: Second line of arguments needs to be lined up.

File: scipy/ndimage/_morphology.py [hunk 4]
```diff
@@ -1365,7 +1365,7 @@ def grey_dilation(input, size=None, footprint=None, structure=None,
         if not sz & 1:
             origin[ii] -= 1
 
-    return filters._min_or_max_filter(input, size, footprint, structure,
+    return _filters._min_or_max_filter(input, size, footprint, structure,
                                       output, mode, cval, origin, 0)
```
Comment (peterbell10) [inline scipy/ndimage/_morphology.py:1369]: Same here

File: scipy/ndimage/_morphology.py (hunk 3 above)
Comment (AnirudhDagar) [reply scipy/ndimage/_morphology.py]: Nice catch! Fixed now.

**Review by rgommers — APPROVED** [review]
Thanks @AnirudhDagar! This looks great. It was a little bit fiddly to get it all right, but I think we've got the right pattern now.

**Pull request merged**